src/Windowing/OpenToolkit.Windowing.Common/Input/KeyboardState.cs

```diff
@@ -17,99 +19,67 @@ namespace OpenToolkit.Windowing.Common.Input
     public struct KeyboardState : IEquatable<KeyboardState>
     {
         // Allocate enough ints to store all keyboard keys
-        private const int IntSize = sizeof(int) * 8;
+        private const int ByteSize = 8;
 
-        private const int NumInts = ((int)Key.LastKey + IntSize - 1) / IntSize;
+        private const int NumBytes = ((int)Key.LastKey + ByteSize) / ByteSize;
```

Comment:
Why the change? The calculation is definitely wrong if we want to use each bit of an int.
E.g. 11 keys would definitely fit into one int:
Old calculation => (11 + 32 - 1) / 32 = 1[.3125]
New calculation => (11 + 8) / 8 = 2[.275]
We could theoretically use bytes instead of ints. But the ints do have a valid reason to be used, namely performance. For 64 Bit it would possibly even make sense to use long instead of ints.

Reply:
I changed the `Key` enum to correctly put `LastKey` as the LAST key instead of last + 1, so I changed this code to account for that.

Also, I'm pretty sure the calculation was off either way (using int just reduces the spots it becomes a problem) but I couldn't find an immediate test case showing the problem so I didn't really bother investigating further.

Does performance actually improve from using ints? I figured it would just increase the struct size.

Reply:
using int at worst case only increases struct size by 24 bits. Which isn't that much considering the size of the struct.
The thing is that an int fits perfectly fine into a register and therefore to check for 32 bits we only need to load the stuff into one register, while with bytes we would need 4 registers(and 4 loads as well). For difference check it's even better as comparison of bytes would need 4 comparisons instead of one as well.
These performance advantages are used by MS as well inside the .net framework and even use it to compare strings, as comparing 64bit at once(they use long) is easier than comparing 8 bytes

Reply:
That's only helpful for checking equality between keyboard states though, isn't it? That's a very rare operation so I wouldn't say it's that significant. The actual bit lookup code is like one instruction different.

I don't see how caches would benefit from ints, if anything they would be hampered because of the extra padding wasting space (but realistically that's insignificant).

That said, I checked in SharpLab, and the JIT code for copying a keyboard state definitely seems more compact with ints.

[SharpLab link](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdzJExQMAsuQAUASmOnzHgBoMEDALwMACYwAGbYHJIYGtQeHu6x9p6OCM7RHgC+8RZM5NbEtnakLm4xCQwAmgwAnv5BoeGRaWVZHnbljlWpWZmlZlkA2nYwGAAWEIEAkuKSjkOj41NikgDyYnwQXLgsAHIQE1ySvFxHAObOALpZOjl5BUneKSVlDN1ZA3Njk9Ozw5+LK2teBstrt9odjlwzpdetkrDZ7O1Kp0ns0YT0elpdNxcNgQjAGLgMFAOGAMAxPDQTDCxFBeApsBh8ZBNmSjmSACrYKAnYYAIV4GFwtXIdDoTTMNLpDKZwNZPAYvKqjNkvAAXviAgAOcUMSX0xkMZmEhhshjbDj4RWMoUBRyc7l8gVCuAMciuAD0CqVMBV6oYAGpXTqaABIa7Y3H4kK8BAwQIMYDehjYfrmy3e3DnYMw8ObSMJiAQSQMADijowjlNwAFziyVNiIfpUAT3uWIRCuGGtWrGHdVp9apgOo8TYTArbHa7AR7DAApF7lYPh+YWuZiAB2BiOFOJxkTzsYc4MABkW/IDAAPBexxh98NnK4AIQBMWvagY2hYvN4glEklk8pKSyPVpUNWUTXle0eQwflBWFUUdRAg0jTlMl+19DUGG0UhENpfUZRZCCyTTfsbS3KDy2dV0PQXAc/UDchszDL8cR/aNY3jU0UxIjMszeHMWPzYBC2LMsYIFSt5R7WsYXrDxGy5Fs93bA9uwFPtvQw5czFHHs7zJacBTnWitNXPoYQ8Dctx3VsVOGI9T0cc8rxvfSHwYZ8GFfNEaHSIA)

Reply:
there would be no padding with ints though? The whole 32-bit of the int would be used.

Next thing is that multiple GetBit can profit from that as well. And by the way does comparing between keystates not happen that unfrequently if you want to know what changed and therefore what happened...

Reply:
When I said padding I also meant the extra unused space from using an int.

> Next thing is that multiple GetBit can profit from that as well.

I don't see how it would. The only difference is a different mov instruction is used between the two.

> And by the way does comparing between keystates not happen that unfrequently if you want to know what changed and therefore what happened...

Yeah that kind of behavior isn't implemented by looking at the different bits and then acting, it's done specifically per key. Ideally if you need this you should just be using the events.

Reply:
> When I said padding I also meant the extra unused space from using an int.
Well the extra spaced used is only at the end, and therefore at max 24bit, which imho is negligeble

> I don't see how it would. The only difference is a different mov instruction is used between the two.
Well forgot to say that we'd need aggressive inlining, before that we won't probably get any advantage

> Yeah that kind of behavior isn't implemented by looking at the different bits and then acting, it's done specifically per key. Ideally if you need this you should just be using the events.

Yeah that's mostly true. But often a polling approach is used and for the ones who do I think we should make it as good as possible(I mean we do not have any big disadvantage with ints, except for the max 24bit padding - right?)
Btw. BitArray uses int arrays as well ;)

Reply:
I changed the implementation back to ints anyways because the JIT generates better code for it while moving them (for some reason..).

Reply:
Well the JIT is stupid. It's not a native compiler, and they did not try to optimize it much at the beginning, they just started doing that. But it will still take a few years till it is really good - I reckon